### PR TITLE
simple digit recognition

### PR DESCRIPTION
**Pull request by robertjankowski**
Demo of digits recognition task. I use pretrained model on MNIST dataset.

### CODE REVIEW

**Comment by robertjankowski** [comment]
Simple UI
ui

**Comment by LaCuneta** [comment]
Thank you for the pull request!  We appreciate it.  We think this is a good example model for using NetLogo and Python together, and we have a few comments for you.

One item is on the Info Tab, which at the moment is mostly empty.  It'd be good to fill in as much detail as you were able to.  You can look at other example models in the NetLogo library for how we usually fill them out.  In particular with this model, it'd be important to note the Python libraries used and what they're for.

There were also some code changes:

- When running a prediction, the output widget area retains the last prediction, which was confusing for me.  You could use `clear-output` to make sure only the most recent prediction is presented, or switch to having a monitor widget for a prediction variable instead of an output area.
- Along those lines, you could have the prediction constantly updating as the drawing changes using a prediction reporter procedure, instead of need to hit a separate button for it.  Any time the `draw` procedure changes the view, just run the prediction.
- One other change is to the `draw` procedure - you can either change the model view updates to be `Continuous` instead of `On Ticks`, or you can add the `display` command to the end of your the procedure (preferred).  Either way, you should be able to remove the `tick` there, since we're not using ticks for anything in this model.
- Finally, you could switch to using an in-memory transfer of the view data instead of using a file as an intermediate store.  Below is a sample of how that might look (warning, I have not tested it).  It just gets the `pcolor` of each patch into a 2D list, which is converted into an array by `np` on the Python side:

```
to-report predict
  py:set images (list map (range max-pycor (min-pycor - 1) -1) [ row ->
    map (range min-pxcor (max-pxcor + 1)) [ col ->
      [ pcolor / 10 ] of patch row col
    ]
  ])
  report py:runresult "model.predict(np.array(images))[0]"
end
```

Thanks, again!

**Comment by robertjankowski** [comment]
ui

Changes:
- automatically predict digit and display on monitor widget
- remote `tick` -> add `display` command
- not saving patch view to file

File: demos/digit-prediction/digits_prediction.nlogo [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+  (py:run
+    "import numpy as np"
+    "from keras.models import load_model"
+  )
+  ;; load model
+  py:run "model = load_model(\"model.h5\")"
+
+  ;; create coords list, useful to make color_list
+  set coords_x (range max-pxcor)
+  set coords_y (range max-pycor)
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: I believe these will miss the last column/row, since `range` is exclusive on the end. They should be, e.g., `range (max-pxcor + 1)`. Also, you might want to start them at (e.g.) `min-pxcor` to be more robust to world sizes. It looks to me like the y-range is flipped, since the y-coordinate is the reverse of row numbers. Is it correctly distinguishing 6s from 9s right now? (can't test ATM) I believe it should be `(range max-pycor (min-pycor - 1) -1)`, unless I'm missing something.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 2]
```diff
@@ -0,0 +1,527 @@
+extensions [ py ]
+
+globals [ path coords_x coords_y color_list digit]
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: In NetLogo, we prefer `-` to `_` in variable names.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 3]
```diff
@@ -0,0 +1,527 @@
...
+  set color_list []
+  foreach coords_x
+  [
+    x -> foreach coords_y
+    [
+      y ->
+        let c [ pcolor / 10 ] of patch x y
+        set color_list lput c color_list
+    ]
+  ]
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: The above would be simplified with `map`:

```
py:set "images" map coords_y [ y ->
  map coords_x  [ x ->
    [ pcolor / 10 ] of patch x y
  ]
]
```

Then there's no need for `reshape` in the python. Just do `array = np.array([images])`.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 4]
```diff
@@ -0,0 +1,527 @@
...
+1
+1
+0
+1
+1
+1
+0
+28
+0
+28
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: I believe your max coordinates on your view should 27, not 28. The max coordinates are inclusive, so min = 0, max = 27 will give you 28 patches across.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+      y ->
+        let c [ pcolor / 10 ] of patch x y
+        set color_list lput c color_list
+    ]
+  ]
+  py:set "images" color_list
+  py:run "array = np.array(images).reshape(28, 28)"
+  py:run "array = array.reshape(1, 28, 28, 1)"
+  py:run "y_pred = model.predict(array)[0]"
+  py:run "result = np.argmax(y_pred)"
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: It could be cool to show the probabilities themselves (I'm assuming the last layer is a softmax). You could even show the probabilities for each of the digits in different monitors. Not necessary, just could be cool.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+  set digit py:runresult "result"
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: You can skip the variable `result` here and just do `py:runresult "np.argmax(y_pred)"`. You could also just do `py:runresult "y_pred"` if you do end up wanting to show the probabilities.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+end
+
+to clear
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: Use the built in `clear-patches` instead.

File: demos/digit-prediction/digits_prediction.nlogo (hunk 2 above)
Comment (qiemem) [reply demos/digit-prediction/digits_prediction.nlogo]: `path` is an unused variable.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+  set coords_y (range max-pycor)
+  set color_list []
+end
+
+to draw
+  ifelse mouse-down?
+  [
+    ask patch mouse-xcor mouse-ycor
+    [
+      ifelse rubber
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: I would rename `rubber` as `erase?`.

File: demos/digit-prediction/digits_prediction.nlogo (hunk 4 above)
Comment (robertjankowski) [reply demos/digit-prediction/digits_prediction.nlogo]: Changed max coords to 27 patches.

File: demos/digit-prediction/digits_prediction.nlogo (hunk 7 above)
Comment (robertjankowski) [reply demos/digit-prediction/digits_prediction.nlogo]: Sure, I didn't know this function.

File: demos/digit-prediction/digits_prediction.nlogo (hunk 3 above)
Comment (robertjankowski) [reply demos/digit-prediction/digits_prediction.nlogo]: Unfortunately I got an error when I pasted this code. 

> MAP expected this input to be a list, but got an anonyous reporter instead

So I didn't change this part of code.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 9]
```diff
@@ -0,0 +1,648 @@
+extensions [ py ]
+
+globals [
+  coords-x coords-y color-list digit
+  zero one two three four five six seven eight nine
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: Rather than storing each probability in it's own variable, you could just store the all the probabilities in a list. I'll show code below.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 10]
```diff
@@ -0,0 +1,648 @@
...
+  py:setup py:python
+  (py:run
+    "import numpy as np"
+    "from keras.models import load_model"
+  )
+  ;; load model
+  py:run "model = load_model(\"model.h5\")"
+
+  ;; create coords list, useful to make color_list
+  set coords-x (range max-pxcor (min-pxcor - 1) -1)
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: `coords-x` still needs to be `(range min-pxcor (max-pxcor + 1))`. This is because is column indices have the same directionality as x cartesian coordinates, but row indices are have opposite directionality as y cartesian coordinates. You'll notice, right now, 5s and 2s are confused, 7s are generally not recognized, and so forth. This will fix that.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 11]
```diff
@@ -0,0 +1,648 @@
...
+      [
+        set pcolor black
+      ]
+      [
+        set pcolor white
+      ]
+    ]
+  ]
+  [
+    predict
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: This is calling `predict` whenever the user isn't drawing (that is, `draw` is pressed, but the user isn't actually doing anything). That's a lot of unnecessary calls. You might consider instead putting the `predict` right after the `ask` instead. This will give users real-time predictions, and use significantly less CPU when idling.

File: demos/digit-prediction/digits_prediction.nlogo (hunk 3 above)
Comment (qiemem) [reply demos/digit-prediction/digits_prediction.nlogo]: Sorry about that! Here's the corrected code:
```
  py:set "images" map [ y ->
    map [ x ->
      [ pcolor / 10 ] of patch x y
    ] coords-x
  ] coords-y
```

File: demos/digit-prediction/digits_prediction.nlogo [hunk 12]
```diff
@@ -0,0 +1,648 @@
...
+  set zero   py:runresult "y_pred[0] * 100"
+  set one    py:runresult "y_pred[1] * 100"
+  set two    py:runresult "y_pred[2] * 100"
+  set three  py:runresult "y_pred[3] * 100"
+  set four   py:runresult "y_pred[4] * 100"
+  set five   py:runresult "y_pred[5] * 100"
+  set six    py:runresult "y_pred[6] * 100"
+  set seven  py:runresult "y_pred[7] * 100"
+  set eight  py:runresult "y_pred[8] * 100"
+  set nine   py:runresult "y_pred[9] * 100"
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: I would just do:

```
  set probabilities py:runresult "y_pred"
```

where `probabilities` is a global. Then, in your monitors, you do (for instance) `100 * item 3 probabilities`. You'd also need to zero out the `probabilities` list in `setup-py`: `set probabilities n-values 10 [ 0 ]`.

File: demos/digit-prediction/digits_prediction.nlogo [hunk 13]
```diff
@@ -0,0 +1,648 @@
...
+end
+
+to clear
```
Comment (qiemem) [inline demos/digit-prediction/digits_prediction.nlogo]: I think this should call `predict`, right?

**Comment by qiemem** [comment]
Oh, one more quick comment (feel free to ignore). You might consider swapping out the NN model for a fully convolutional one (so, no `Dense` layers). Right now, performance suffers when the digit is too far left or right. For instance, try drawing a straight line near the right or left side. It will register as a 4 or a 6 instead of a 1. A fully convolutional network should fix that. Maybe something like adding another conv layer or two on and then finish up with a global avg pooling layer with a softmax activation.

**Comment by robertjankowski** [comment]
I just added fixes. Referring to the model I think maybe in couple of days I would try to improve NN architecture. Many thanks for your help!

File: demos/digit-prediction/digits_prediction.nlogo [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+
+  set probabilities n-values 10 [ 0 ]
+end
+
+to draw
+  if mouse-down?
+  [
+    ask patch mouse-xcor mouse-ycor
+    [
+      predict
```
[the remark this answers is not in the document]
Comment (robertjankowski) [reply demos/digit-prediction/digits_prediction.nlogo]: Good point!

**Comment by LaCuneta** [comment]
This looks good to me.  @qiemem, @robertjankowski: Any further changes, or are we good to merge?

**Comment by robertjankowski** [comment]
For me PR is ready to merge.
